exasol_transformers_extension/udfs/sequence_classification_single_text_udf.py

```diff
@@ -7,28 +8,70 @@ class SequenceClassificationSingleText:
     def __init__(self,
                  exa,
                  cache_dir=None,
+                 batch_size=100,
                  base_model=transformers.AutoModelForSequenceClassification,
                  tokenizer=transformers.AutoTokenizer):
         self.exa = exa
         self.cache_dir = cache_dir
+        self.bacth_size = batch_size
         self.base_model = base_model
         self.tokenizer = tokenizer
+        self.last_loaded_model_name = None
+        self.last_loaded_model = None
+        self.last_loaded_tokenizer = None
 
     def run(self, ctx):
-        bucketfs_conn = ctx.bucketfs_conn
-        text_data = ctx.text_data
-        model_name = ctx.model_name
+        while True:
+            batch_df = ctx.get_dataframe(self.bacth_size)
+            if batch_df is None:
+                break
+
+            result_df = self.get_batched_predictions(batch_df)
+            ctx.emit(result_df)
+
+    def get_batched_predictions(self, batch_df):
+        result_df_list = []
+        for model_name in batch_df['model_name'].unique():
+            model_df = batch_df[batch_df['model_name'] == model_name]
+            bucketfs_conn = model_df['bucketfs_conn'].iloc[0]
+
+            if self.last_loaded_model_name != model_name:
+                self.load_models(model_name, bucketfs_conn)
+
+            model_pred_df = self.model_prediction(model_df)
+            result_df_list.append(model_pred_df)
+
+        result_df = pd.concat(result_df_list)
+        return result_df
+
+    def load_models(self, model_name, bucketfs_conn):
         self._set_cache_directory(model_name, bucketfs_conn)
 
-        model = self.base_model.from_pretrained(
+        self.last_loaded_model = self.base_model.from_pretrained(
             model_name, cache_dir=self.cache_dir)
-        tokenizer = self.tokenizer.from_pretrained(
+        self.last_loaded_tokenizer = self.tokenizer.from_pretrained(
             model_name, cache_dir=self.cache_dir)
+        self.last_loaded_model_name = model_name
+
+    def model_prediction(self, model_df):
+        bucketfs_conn = model_df['bucketfs_conn'].iloc[0]
+        model_name = model_df['model_name'].iloc[0]
+
+        model_preds = []
+        for ix, row in model_df.iterrows():
+            text_data = row['text_data']
+            tokens = self.last_loaded_tokenizer(text_data, return_tensors="pt")
+            logits = self.last_loaded_model(**tokens).logits
```

Comment:
can you check, if tokenizer and/or model have batch capabilities. Here a first hint https://huggingface.co/course/chapter2/5?fw=tf. Another thing, for later, maybe think about, how you can refactor this into using two Threads, one thread which interacts with the UDF context and one who does the prediction. The reason for this refactoring is performance. The prediction will run eventually on GPU and during that the CPU is free to request new data from the DB or emit results.

Reply:
I thought about it and even tested the implementation on the same page. We may need to do a few workarounds, should I create a refactroring ticket without make this PR too complicated?

PredictionUDF is already decoupled from DownloaderUDF. So, two thread mechanism will be only useful for performing prediction on the GPU (as you said) and emitting results on CPU. We can investigate this in another ticket